### PR TITLE
Change IOutput to merge methods and return Memory

### PR DESCRIPTION
**Pull request by pakrym**
`Memory<byte> GetMemory()` for now.

### CODE REVIEW

File: samples/LowAllocationWebServer/LowAllocationWebServerLibrary/TcpConnectionFormatter.cs [hunk 1]
```diff
@@ -30,23 +30,21 @@ public TcpConnectionFormatter(TcpConnection connection, int bufferSize = 4096)
...
-        }
-
         public void Advance(int bytes)
         {
             _written += bytes;
             if (_written >= _buffer.Length) throw new InvalidOperationException();
         }
 
-        public void Enlarge(int desiredBufferLength = 0)
+        public Memory<byte> GetMemory(int desiredBufferLength = 0)
```
Comment (KrzysztofCwalina) [inline samples/LowAllocationWebServer/LowAllocationWebServerLibrary/TcpConnectionFormatter.cs]: Can you change the parameters to minimumLength?

File: src/System.Buffers.Primitives/System/Buffers/IOutput.cs [hunk 2]
```diff
@@ -5,10 +5,9 @@ namespace System.Buffers
 {
     public interface IOutput
     {
-        Span<byte> GetSpan();
         void Advance(int bytes);
 
         /// <summary>desiredBufferLength == 0 means "i don't care"</summary>
-        void Enlarge(int desiredBufferLength = 0);
+        Memory<byte> GetMemory(int desiredBufferLength = 0);
```
Comment (KrzysztofCwalina) [inline src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: Why are we changing it from Span to Memory?

Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: Also, let's rename the parameter to minimumLength.

Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: Also, let's describe the contract in the doc comment.

File: src/System.Text.Formatting/System/Text/Formatting/IOutputExtensions.cs [hunk 3]
```diff
@@ -12,13 +12,13 @@ public static class IOutputExtensions
         public static void Append<TFormatter, T>(this TFormatter formatter, T value, SymbolTable symbolTable, StandardFormat format = default) where T : IBufferFormattable where TFormatter : IOutput
         {
             while (!formatter.TryAppend(value, symbolTable, format)) {
-                formatter.Enlarge();
+                formatter.GetMemory();
```
Comment (KrzysztofCwalina) [inline src/System.Text.Formatting/System/Text/Formatting/IOutputExtensions.cs]: This needs to force enlarge. It cannot simply call GetMemory.

File: src/System.Buffers.Primitives/System/Buffers/IOutput.cs (hunk 2 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/IOutput.cs]: It was in the email:

1.	It aligns with ROB that has memory and Reader/Writer that operate on span
2.	Allows IOuput to be used with async methods (scenario: get memory from IOutput, pass to Stread.ReadAsync)

File: src/System.Text.Formatting/System/Text/Formatting/IOutputExtensions.cs [hunk 4]
```diff
@@ -12,13 +12,13 @@ public static class IOutputExtensions
...
             }
         }
 
         public static bool TryAppend<TFormatter, T>(this TFormatter formatter, T value, SymbolTable symbolTable, StandardFormat format = default) where T : IBufferFormattable where TFormatter : IOutput
         {
-            if (!value.TryFormat(formatter.GetSpan(), out int bytesWritten, format, symbolTable))
+            if (!value.TryFormat(formatter.GetMemory().Span, out int bytesWritten, format, symbolTable))
```
Comment (KrzysztofCwalina) [inline src/System.Text.Formatting/System/Text/Formatting/IOutputExtensions.cs]: These .Span calls are annoying.

File: src/System.Text.Formatting/System/Text/Formatting/ITextOutputExtensions.cs [hunk 5]
```diff
@@ -303,18 +303,23 @@ public static class ITextOutputExtensions
...
             {
                 return false;
             }
             formatter.Advance(bytesWritten);
             return true;
         }
+
+        private static void GetMoreMemory<TFormatter>(this TFormatter formatter) where TFormatter : ITextOutput
+        {
+            formatter.GetMemory(formatter.GetMemory().Length * 2);
```
Comment (KrzysztofCwalina) [inline src/System.Text.Formatting/System/Text/Formatting/ITextOutputExtensions.cs]: Yeah, this looks inefficient.

Comment (pakrym) [reply src/System.Text.Formatting/System/Text/Formatting/ITextOutputExtensions.cs]: These API's have to be on OuputWriter anyway.

File: src/System.Text.Formatting/System/Text/Formatting/IOutputExtensions.cs (hunk 4 above)
Comment (pakrym) [reply src/System.Text.Formatting/System/Text/Formatting/IOutputExtensions.cs]: There would be no .Span class if we move to OutputWriter

**Comment by pakrym** [comment]
Added `GetSpan`, renamed arguments, ready for review.

File: src/System.Text.Formatting/System/Text/Formatting/Formatters/StreamFormatter.cs [hunk 6]
```diff
@@ -31,15 +31,16 @@ public StreamFormatter(Stream stream, SymbolTable symbolTable, ArrayPool<byte> p
         }
         Memory<byte> IOutput.GetMemory(int minimumLength)
         {
-            var newSize = _buffer.Length * 2;
-            if(minimumLength != 0){
-                newSize = minimumLength;
+            if (minimumLength > _buffer.Length)
+            {
+                var newSize = _buffer.Length * 2;
```
Comment (ahsonkhan) [inline src/System.Text.Formatting/System/Text/Formatting/Formatters/StreamFormatter.cs]: nit:
`newSize = minimumLength == 0 ? _buffer.Length * 2 : minimumLength;`

File: src/System.Text.Formatting/System/Text/Formatting/Formatters/StreamFormatter.cs [hunk 7]
```diff
@@ -31,15 +31,16 @@ public StreamFormatter(Stream stream, SymbolTable symbolTable, ArrayPool<byte> p
...
+                if(minimumLength != 0){
+                    newSize = minimumLength;
+                }
+                var temp = _buffer;
```
Comment (ahsonkhan) [inline src/System.Text.Formatting/System/Text/Formatting/Formatters/StreamFormatter.cs]: nit: specify type here

File: src/System.Text.Formatting/System/Text/Formatting/Formatters/ArrayFormatter.cs [hunk 8]
```diff
@@ -32,18 +32,22 @@ public ArrayFormatter(int capacity, SymbolTable symbolTable, ArrayPool<byte> poo
...
-            var doubleCount = _buffer.Free.Count * 2;
-            int newSize = desiredBufferLength>doubleCount?desiredBufferLength:doubleCount;
-            var newArray = _pool.Rent(newSize + _buffer.Count);
-            var oldArray = _buffer.Resize(newArray);
-            _pool.Return(oldArray);
+            if (minimumLength < 1) minimumLength = 1;
+            if (minimumLength > _buffer.Free.Count)
+            {
+                var doubleCount = _buffer.Free.Count * 2;
+                int newSize = minimumLength>doubleCount?minimumLength:doubleCount;
```
Comment (ahsonkhan) [inline src/System.Text.Formatting/System/Text/Formatting/Formatters/ArrayFormatter.cs]: nit: add spacing

**Comment by pakrym** [comment]
This PR was not intended to improve code style in experimental code. I left it as is. We can file a separate issue to reformat the code.

**Comment by pakrym** [comment]
Any objection to merging this?

@KrzysztofCwalina @davidfowl @joshfree @muratg

**Comment by KrzysztofCwalina** [comment]
Looks good. Thanks!

**Pull request merged**